### PR TITLE
added exception handling for latest release checking and tests

### PR DESCRIPTION
**Pull request by angus-yxz**
Gitmastery attempts to check for latest release via a requests get to github. This get has no exception handling so any connection errors either locally or remotely will cause the entire application to crash when the right thing to do is to handle and log the exception and continue on. 

I added a small fix to handle this exception, logging a warning when the release check fails. Additionally, I added a timeout because on some networking configurations the CLI might just hang. I added a single test which sets env vars (that python's requests library respects https://requests.readthedocs.io/en/latest/user/advanced/#proxies) to mock a connection failure. Also ruff check shows formatting errors that have nothing to do with my own diffs, they were present in older commits so I just left those alone and formatted my own code to minimize the diff.

Also I know you asked in CONTRIBUTING.md for me to open an issue to ask to have this be assigned to me explicitly but its 5:30 in the morning right now and the change was small and focused enough to make on my own quickly to have a locally working binary so I didnt get stuck waiting.

### CODE REVIEW

File: app/cli.py [hunk 1]
```diff
@@ -23,6 +23,9 @@ def invoke(self, ctx: click.Context) -> None:
 
 CONTEXT_SETTINGS = {"max_content_width": 120}
 
+# Bound the release check so the CLI cannot hang on an unresponsive network
+LATEST_RELEASE_TIMEOUT_SECONDS = 5
```
Comment (VikramGoyal23) [inline app/cli.py]: Timeout is too small; preferably wait 20 seconds at least

File: app/cli.py [hunk 2]
```diff
@@ -39,14 +42,25 @@ def cli(ctx: click.Context, verbose: bool) -> None:
...
+        # there is no version to compare against
+        location = response.headers.get("Location")
+        if location is not None:
+            latest_version = Version.parse_version_string(location.rsplit("/", 1)[-1])
+    except (requests.exceptions.RequestException, ValueError):
+        latest_version = None
+
+    if latest_version is None:
+        warn("Unable to verify the latest version release")
+    elif current_version.is_behind(latest_version):
```
Comment (VikramGoyal23) [inline app/cli.py]: Ought to still bubble up the actual exception that occured for extra information

**Review by VikramGoyal23 — DISMISSED** [review]
Hi! Thanks so much for your contribution! There's a couple of issues with the implementation, but it's a good step in the right direction!

**Comment by jovnc** [comment]
@angus-yxz Thanks for your contribution and user feedback!

Generally, the implementation is ok, however I think a better direction for us would be a global exception handler, rather than a specific scenario exception handler. We will consider this in future improvements of gitmastery.

For this case, whether to allow users to continue after failure to check latest release, we can consider as well. Likely case is that offline users who cannot fetch latest release would not be able to use `gitmastery` as well, as our exercises download and verification requires internet connection.

cc: @VikramGoyal23

File: app/cli.py (hunk 1 above)
Comment (jovnc) [reply app/cli.py]: In terms of UX, 5s should be sufficient for timeout checks, since its non-critical anyways. If we set too long, offline users might wait 20s before the commands continues execution.

File: app/cli.py (hunk 2 above)
Comment (jovnc) [reply app/cli.py]: Raising actual exception would fail request early.

I'm unsure whether latest update checking should be part of the critical path yet, but as of now I think it makes sense to not be part of critical path.
* If yes, should fail early
* If not, we can re-log the actual exception instead of `raise`

**Comment by angus-yxz** [comment]
> @angus-yxz Thanks for your contribution and user feedback!
> 
> Generally, the implementation is ok, however I think a better direction for us would be a global exception handler, rather than a specific scenario exception handler. We will consider this in future improvements of gitmastery.
> 
> For this case, whether to allow users to continue after failure to check latest release, we can consider as well. Likely case is that offline users who cannot fetch latest release would not be able to use `gitmastery` as well, as our exercises download and verification requires internet connection.
> 
> cc: @VikramGoyal23

Hi, yes, I agree with that approach, I didnt look too deeply into the source code implementation, and Im also not exactly sure why, but all the exercise download and verification worked on my end  _even when_ the gitmastery release check failed via that http get call. I dont think a single http release get failure should block the rest of the CLI from operating. In my case this patch was necessary for me to just use the CLI (otherwise I literally would be unable to continue with the rest of the lessons). I think the most idiomatic thing to do in this case is to log the failure and then continue attempting executing the gh commands and if those in turn fail _then_ it is appropriate to raise and error out loudly rather than preemptively speculate and gate the rest of the application CLI functionality behind a single failed release check

For the other maintainers, here is the full context behind the change: https://github.com/NUS-CS2103-AY2627-S1/forum/issues/8#issue-5135525611

I am still encountering spurious connection closing on my end and other users have reported similar issues on particular networking setups, so this change is necessary just to access the functionality of the CLI since the gh commands still go through fine. Regardless, I think it is still a best practice to handle this single release check exception properly instead of letting it bubble and crash the entire application preventing usage of the rest of the functionality

File: tests/e2e/test_version.py
Comment (jovnc) [inline tests/e2e/test_version.py:1]: Let's remove this E2E test, currently our test plan is to test only main app function / critical path as E2E tests are expensive to run. For this edge case, its ok to leave it out.

File: app/cli.py [hunk 3]
```diff
@@ -39,14 +42,25 @@ def cli(ctx: click.Context, verbose: bool) -> None:
...
-        .headers["Location"]
-        .rsplit("/", 1)[-1]
-    )
-    if current_version.is_behind(Version.parse_version_string(latest_version)):
+        # GitHub redirects to the tag of the latest release; without the redirect
+        # there is no version to compare against
+        location = response.headers.get("Location")
+        if location is not None:
+            latest_version = Version.parse_version_string(location.rsplit("/", 1)[-1])
+    except (requests.exceptions.RequestException, ValueError):
```
Comment (jovnc) [inline app/cli.py]: Log warning here, with the actual exception message here.

File: app/cli.py [hunk 4]
```diff
@@ -39,14 +42,25 @@ def cli(ctx: click.Context, verbose: bool) -> None:
...
+        latest_version = None
+
+    if latest_version is None:
+        warn("Unable to verify the latest version release")
```
Comment (jovnc) [inline app/cli.py]: No need to warn here if warn above already, then log version behind if `latest_version` is not `None` and `current_version.is_behind(latest_version):`

**Review by jovnc — COMMENTED** [review]
Yup, agree with you that latest version check should not be in critical path, just some small code changes and should be ok to merge, thanks @angus-yxz !

**Comment by angus-yxz** [comment]
accidentally pushed the commit with author attribution to my personal account (forgot to overwrite my global config) force pushed to change attribution back

File: tests/e2e/test_version.py
Comment (angus-yxz) [reply tests/e2e/test_version.py:1]: Done

File: app/cli.py (hunk 3 above)
Comment (angus-yxz) [reply app/cli.py]: Done + added more granular logging of error conditions

File: app/cli.py (hunk 4 above)
Comment (angus-yxz) [reply app/cli.py]: Done

File: app/cli.py (hunk 2 above)
Comment (angus-yxz) [reply app/cli.py]: Done, consensus reached with @jovnc for handling update checking behaviour

**Review by jovnc — APPROVED** [review]
LGTM updated the code abit to tidy it up, otherwise its good, thanks for the help identifying this issue!

Ignore the E2E test failures, due to pull_request_target issues

**Comment by damithc** [comment]
Good work, @angus-yxz Thanks for your contribution to Git-Mastery.
